### PR TITLE
fix ledger transactions

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/HW/HW.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  // .add('staking registration', () => (
+  //   <ResetWrapper>
+  //     <SignTxWithLedger ledgerPayload={ledgerPayload.staking.registration} />
+  //   </ResetWrapper>
+  // ))
+  // .add('staking deregistration', () => (
+  //   <ResetWrapper>
+  //     <SignTxWithLedger ledgerPayload={ledgerPayload.staking.deregistration} />
+  //   </ResetWrapper>
+  // ))
```
Comment (stackchain) [inline src/HW/HW.stories.tsx]:
```suggestion
```

File: src/HW/HW.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      .catch((error) => {
+        setError(error)
+        setStep(Step.Error)
+      })
+  }
+
+  return (
+    <View>
+      {!!signedLedgerTx && <Text>Signed Ledger Tx: {JSON.stringify(signedLedgerTx, null, 2)}</Text>}
+      {error && <Text>error: {error.message}</Text>}
```
Comment (stackchain) [inline src/HW/HW.stories.tsx]:
```suggestion
      {error != null && <Text>error: {error.message}</Text>}
```

File: src/HW/mocks.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      },
+      additionalWitnessPaths: [],
+    } as SignTransactionRequest,
+  },
+  staking: {
+    // registration: {} as SignTransactionRequest,
+    // deregistration: {} as SignTransactionRequest,
+    // delegation: {} as SignTransactionRequest,
+    // withdrawal: {} as SignTransactionRequest,
+  },
```
Comment (stackchain) [inline src/HW/mocks.ts]:
```suggestion
```

File: src/yoroi-wallets/cardano/unsignedTx.ts [hunk 4]
```diff
@@ -198,3 +204,25 @@ const Voting = {
     votingRegistration?: VotingRegistration
   }): Promise<YoroiVoting['registration']> => votingRegistration,
 }
+
+export const toBech32 = (address: string) => {
+  const baseAddressRegExp = /^[0-3]?[0-9a-f]+$/i
+  const rewardAddressRegExp = /^[ef]?[0-9a-fA-F]+$/i
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/unsignedTx.ts]: Do only reward addresses take uppercase hex?

File: src/yoroi-wallets/cardano/unsignedTx.ts [hunk 5]
```diff
@@ -198,3 +204,25 @@ const Voting = {
     votingRegistration?: VotingRegistration
   }): Promise<YoroiVoting['registration']> => votingRegistration,
 }
+
+export const toBech32 = (address: string) => {
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/unsignedTx.ts]: Can we have a test for this?

**Review by stackchain — CHANGES_REQUESTED** [review]

File: src/yoroi-wallets/cardano/unsignedTx.ts (hunk 5 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/unsignedTx.ts]: nope

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 6]
```diff
@@ -771,33 +769,22 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
...
     const ledgerPayload = await YoroiLib.cardano.buildLedgerPayload(
       unsignedTx.unsignedTx,
-      (this._getBaseNetworkConfig() as any).PROTOCOL_MAGIC,
       Number.parseInt(this._getChainNetworkId(), 10),
-      addressingMap,
+      (this._getBaseNetworkConfig() as any).PROTOCOL_MAGIC,
+      this.getStakingKeyPath(),
     )
 
     const signedLedgerTx = await signTxWithLedger(ledgerPayload, this.hwDeviceInfo, useUSB)
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:781]: signTxWithLedger -> signTxWithLedger

File: src/yoroi-wallets/cardano/unsignedTx.ts [hunk 7]
```diff
@@ -211,33 +211,12 @@ const Voting = {
...
-  return baseAddressRegExp.test(address)
-}
-
-const isRewardAddress = (address: string) => {
-  const rewardAddressRegExp = /^[ef][0-9a-f]+$/i
-
-  return rewardAddressRegExp.test(address)
-}
+const isBaseAddress = (address: string) => ['0', '1', '2', '3'].includes(address.charAt(0))
+const isPointerAddress = (address: string) => ['4', '5'].includes(address.charAt(0))
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/unsignedTx.ts]: `0x[4]1` = `A`, `0x[4]4` = `D`, `A` = `Byron`, `D` = `Dead/Icarus`
It needs to check for byron first, that's the only way.
https://github.com/Emurgo/yoroi-frontend/blob/9a29a38cf69b7d4472a94615c9255101b318c74e/packages/yoroi-extension/app/api/ada/lib/storage/bridge/utils.js#L277

**Review by stackchain — CHANGES_REQUESTED** [review]

File: src/yoroi-wallets/cardano/ShelleyWallet.ts (hunk 6 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:781]: ?

Comment (stackchain) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:781]: shadowing

Comment (wolverineks) [reply src/yoroi-wallets/cardano/ShelleyWallet.ts:781]: ill change it in another pr

**Review by stackchain — APPROVED** [review]

**Pull request merged**